File: tests/fixtures/python/buggy.py
# EXPECTED FINDINGS:
# - line 14: crypto-hardcoded-secret — security — Hardcoded secret literal
# - line 19: injection-sql-f-string — security — cursor.execute with f-string SQL
# - line 22: injection-os-system — security — os.system shell injection
# - line 25: injection-shell-true — security — subprocess.run shell=True
# - line 28: crypto-md5 — security — hashlib.md5 weak hash
# - line 31: crypto-pickle-load — security — pickle.loads on untrusted data
# - line 34: crypto-yaml-load — security — yaml.load without SafeLoader
# - line 37: injection-eval — security — eval on user input
# - line 38: bug-mutable-default-arg — bug — mutable default argument
# - line 44: crypto-weak-random — security — random.random() not cryptographically secure
# - line 47: bug-string-concat-int — bug — implicit string+int concatenation
"""Intentionally buggy Python fixture for DecodeMind scanner tests."""
API_TOKEN = "sk-supersecretvalue1234"

import hashlib, os, pickle, random, subprocess, yaml

def lookup_user(cursor, name):
    cursor.execute(f"SELECT * FROM users WHERE name = '{name}'")

def run_command(filename):
    os.system("cat " + filename)

def run_pipeline(cmd):
    subprocess.run(cmd, shell=True)

def fingerprint(data):
    return hashlib.md5(data).hexdigest()

def load_session(blob):
    return pickle.loads(blob)

def parse_config(text):
    return yaml.load(text)

def calc(expr):
    return eval(expr)
def append_item(item, bucket=[]):
    bucket.append(item)
    return bucket


def token():
    return str(random.random())

def label(prefix):
    return prefix + 5

File: tests/fixtures/typescript/buggy.ts
// EXPECTED FINDINGS:
// - line 16: crypto-hardcoded-secret — security — Hardcoded SECRET literal
// - line 19: injection-js-eval — security — eval() on user input
// - line 23: injection-js-innerhtml — security — innerHTML assignment with variable
// - line 27: xss-document-write — security — document.write with user input
// - line 31: crypto-md5-js — security — crypto.createHash('md5') weak hash
// - line 35: crypto-jwt-none-alg — security — JWT verified with 'none' algorithm
// - line 39: logic-json-parse-no-try — logic — JSON.parse without try/catch
// - line 42: quality-any-type — quality — variable typed any
// - line 46: quality-no-jsdoc — quality — exported function without JSDoc
// - line 50: quality-console-log-prod — quality — console.log in production

import crypto from 'crypto';
import jwt from 'jsonwebtoken';

const SECRET: string = "supersecret-prod-key-2026";

export function calc(expr: string): unknown {
  return eval(expr);
}

export function renderHtml(el: HTMLElement, html: string): void {
  el.innerHTML = html;
}

export function inject(payload: string): void {
  document.write(payload);
}

export function fingerprint(data: string): string {
  return crypto.createHash('md5').update(data).digest('hex');
}

export function verifyToken(token: string, secret: string): unknown {
  return jwt.verify(token, secret, { algorithms: ['none'] });
}

export function parseConfig(raw: string): unknown {
  return JSON.parse(raw);
}

export function legacyHandler(payload: any): void {
  console.error(payload);
}

export function fetchSomething(url: string): Promise<Response> {
  return fetch(url);
}

export function logTrace(message: string): void {
  console.log("trace:", message);
}
